modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/DemuxWorkflowOperationHandler.java

```diff
@@ -0,0 +1,260 @@
+        throw new WorkflowOperationException(String.format("Source flavor '%s' is malformed", flavor));
+      }
+    }
+
+    // Select the source tags
+    for (String tag : asList(sourceTagsOption)) {
+      elementSelector.addTag(tag);
+    }
+
+    // Find the encoding profile - should only be one
+    EncodingProfile profile = composerService.getProfile(encodingProfile);
+    if (profile == null) {
+      throw new WorkflowOperationException(String.format("Encoding profile '%s' was not found", encodingProfile));
+    }
+    // Look for elements matching the tag
+    Collection<Track> sourceTracks = elementSelector.select(mediaPackage, false);
+    if (sourceTracks.isEmpty()) {
+      logger.info("No matching tracks found");
+      return createResult(mediaPackage, Action.CONTINUE);
+    }
+
+    long totalTimeInQueue = 0;
+    Map<Job, Track> encodingJobs = new HashMap<>();
+    for (Track track : sourceTracks) {
+      logger.info("Encoding track {} using encoding profile '{}'", track, profile);
+      // Start encoding and wait for the result
+      encodingJobs.put(composerService.demux(track, profile.getIdentifier()), track);
+    }
+
+    // Wait for the jobs to return
+    if (!waitForStatus(encodingJobs.keySet().toArray(new Job[encodingJobs.size()])).isSuccess()) {
+      throw new WorkflowOperationException("One of the encoding jobs did not complete successfully");
+    }
+
+    // Process the result
+    for (Map.Entry<Job, Track> entry : encodingJobs.entrySet()) {
+      Job job = entry.getKey();
+      Track sourceTrack = entry.getValue();
+
+      // add this receipt's queue time to the total
+      totalTimeInQueue += job.getQueueTime();
+
+      // it is allowed for compose jobs to return an empty payload. See the EncodeEngine interface
+      if (job.getPayload().length() <= 0) {
+        logger.warn("No output from Demux operation");
+        continue;
+      }
+
+      List<Track> composedTracks = (List<Track>) MediaPackageElementParser.getArrayFromXml(job.getPayload());
+      if (composedTracks.size() != targetFlavors.size() && targetFlavors.size() != 1) {
+        throw new WorkflowOperationException(String.format("Number of target flavors (%d) and output tracks (%d) do "
+                + "not match", targetFlavors.size(), composedTracks.size()));
+      }
+      if (composedTracks.size() != targetTags.length && targetTags.length != 1 && targetTags.length != 0) {
+        throw new WorkflowOperationException(String.format("Number of target tag groups (%d) and output tracks (%d) "
+                + "do not match", targetTags.length, composedTracks.size()));
+      }
+
+      // Flavor each track in the order read
+      int i = 0;
+      int j = 0;
+      for (Track composedTrack : composedTracks) {
+        // set flavor to the matching flavor in the order listed
+        composedTrack.setFlavor(newFlavor(sourceTrack, targetFlavors.get(i)));
+        if (targetFlavors.size() > 1) {
```

Comment:
Somewhat weird to read as based on the fact that at most to composedTracks are supported which is not obvious if you only look at this loop.